### PR TITLE
fix sig validation to accept and ignore the FORKID bit prior to the fork

### PR DESCRIPTION
**Pull request by gandrewstone**

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -4321,8 +4321,9 @@ bool ProcessNewBlock(CValidationState &state,
     if (!checked)
     {
         int byteLen = ::GetSerializeSize(*pblock, SER_NETWORK, PROTOCOL_VERSION);
-        LogPrintf("Invalid block: ver:%x time:%d Tx size:%d len:%d\n", pblock->nVersion, pblock->nTime,
-            pblock->vtx.size(), byteLen);
+        LogPrintf("Invalid block: ver:%x time:%d Tx size:%d len:%d reject code: 0x%x reject reason: %s debug: %s \n",
```
Comment (sickpig) [inline src/main.cpp]: too verbose?

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -1276,8 +1255,18 @@ bool TransactionSignatureChecker::CheckSig(const vector<unsigned char>& vchSigIn
     int nHashType = vchSig.back();
     vchSig.pop_back();
 
+    uint256 sighash;
     size_t nHashed = 0;
-    uint256 sighash = SignatureHash(scriptCode, *txTo, nIn, nHashType, amount, &nHashed);
+    // If BCC sighash is possible, check the bit, otherwise ignore the bit.  This is needed because
+    // the bit is undefined (can be any value) before the fork. See block 264084 tx 102
+    if ((nFlags & SCRIPT_ENABLE_SIGHASH_FORKID)&&(nHashType & SIGHASH_FORKID))
```
Comment (sickpig) [inline src/script/interpreter.cpp:1271]: nit: space between parenthesis and logical operator

File: src/script/sign.cpp [hunk 3]
```diff
@@ -19,15 +19,17 @@ using namespace std;
...
 bool TransactionSignatureCreator::CreateSig(std::vector<unsigned char>& vchSig, const CKeyID& address, const CScript& scriptCode) const
 {
     CKey key;
     if (!keystore->GetKey(address, key))
         return false;
 
-    uint256 hash = SignatureHash(scriptCode, *txTo, nIn, nHashType, amount);
+    uint256 hash;
+    if (nHashType & SIGHASH_FORKID) hash = SignatureHashBitcoinCash(scriptCode, *txTo, nIn, nHashType, amount);
+    else hash = SignatureHash(scriptCode, *txTo, nIn, nHashType, amount);
```
Comment (sickpig) [inline src/script/sign.cpp]: nit: if statement formatting

or alternatively what about 

```c++
hash = SignatureHash(scriptCode, *txTo, nIn, nHashType, amount);
if (nHashType & SIGHASH_FORKID) 
       hash = SignatureHashBitcoinCash(scriptCode, *txTo, nIn, nHashType, amount);
```

or use the same style you apply to https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/740/files#diff-a0ec10dadca1c87cc43ef3113574a264L317

File: src/test/script_tests.cpp [hunk 4]
```diff
@@ -154,7 +153,17 @@ void DoTest(const CScript& scriptPubKey, const CScript& scriptSig, int flags, co
     CMutableTransaction txCredit = BuildCreditingTransaction(scriptPubKey, nValue);
     CMutableTransaction tx = BuildSpendingTransaction(scriptSig, txCredit);
     CMutableTransaction tx2 = tx;
-    BOOST_CHECK_MESSAGE(VerifyScript(scriptSig, scriptPubKey, flags, MutableTransactionSignatureChecker(&tx, 0, txCredit.vout[0].nValue), &err) == expect, message);
+    if (1) //flags & SIGHASH_FORKID)
```
Comment (sickpig) [inline src/test/script_tests.cpp]: why `if(1)`, wouldn't it be better to add a comment  and to keep trace of what was before?

File: src/test/script_tests.cpp [hunk 5]
```diff
@@ -154,7 +153,17 @@ void DoTest(const CScript& scriptPubKey, const CScript& scriptSig, int flags, co
...
+    {
+        bool result = VerifyScript(scriptSig, scriptPubKey, flags, MutableTransactionSignatureChecker(&tx, 0, txCredit.vout[0].nValue, flags), &err);
+        if (result != expect)
+        {
+            printf("bug\n");
```
Comment (sickpig) [inline src/test/script_tests.cpp]: debug left over.
Using `DebugAssert()` instead?

File: src/test/txvalidationcache_tests.cpp [hunk 6]
```diff
@@ -54,7 +54,15 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup)
...
-        uint256 hash = SignatureHash(scriptPubKey, spends[i], 0, sighashType, coinbaseTxns[0].vout[0].nValue, 0);
+        uint256 hash;
+        if (sighashType & SIGHASH_FORKID)
+        {
+        hash = SignatureHashBitcoinCash(scriptPubKey, spends[i], 0, sighashType, coinbaseTxns[0].vout[0].nValue, 0);
+        }
+        else
+        {
+        hash = SignatureHash(scriptPubKey, spends[i], 0, sighashType, coinbaseTxns[0].vout[0].nValue, 0);
+        }
```
Comment (sickpig) [inline src/test/txvalidationcache_tests.cpp]: indentation/formatting

**Comment by sickpig** [comment]
failed with 
```
Running testscript bip68-112-113-p2p.py ...
Random seed: 1502293585
Initializing test directory  /tmp/testw1f2aha_
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
MiniNode: Connecting to Bitcoin Node IP # 127.0.0.1:11211
Assertion failed: Not all nodes requested block
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/test_framework.py", line 186, in main
    self.run_test()
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/build/../qa/rpc-tests/bip68-112-113-p2p.py", line 110, in run_test
    test.run()
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/comptool.py", line 370, in run
    self.sync_blocks(block.sha256, 1)
  File "/home/travis/build/BitcoinUnlimited/BitcoinUnlimited/qa/rpc-tests/test_framework/comptool.py", line 211, in sync_blocks
    raise AssertionError("Not all nodes requested block")
Stopping nodes
```
kicking travis job

**Review by ptschip — APPROVED** [review]

**Pull request merged**